src/redprl/refiner.fun

```diff
@@ -970,49 +970,46 @@ struct
       end
   end
 
-  structure Restriction =
-  struct
-    (* This structure provides functions that automate the restriction
-       judgement rules given in "Dependent Cubical Realizability",
-       page 46.
+  structure Restriction :
+  sig
+    (* This structure used to provide functions that automate the
+       restriction judgement rules given in "Dependent Cubical
+       Realizability", page 46.
+
+       On 2017/06/14, favonia implemented a function to handle
+       all cases. We need to generalize the old handling to
+       welcome new hcom operators. (It is true that we still
+       only need to handle context restrictions with at most two
+       equations, but it seems easier to just handle all cases.)
 
        Such automation is possible because the rules are "syntax
        directed" on the restriction being performed.
      *)
 
-    (* Return whether given dimensions are different constants, which
-       is a contradiction, allowing the restriction judgment to conclude
-       anything. *)
-    fun dimsContradictory (P.APP P.DIM0, P.APP DIM1) = true
-      | dimsContradictory (P.APP P.DIM1, P.APP DIM0) = true
-      | dimsContradictory _ = false
-
-    (* Restrict a judgement by a single equation `ext = eps`. *)
-    fun One (jdg : abt jdg) (ext : param) (eps : param) : abt jdg list =
-      if P.eq Sym.eq (ext, eps) then [jdg] (* combining rules in first row *)
-      else if dimsContradictory (ext, eps) then [] (* second row, left rule *)
-      else
-        let
-          val (P.VAR v, P.APP eps) = (ext, eps)
-        in
-          [Seq.map (substSymbol (P.APP eps, v)) jdg] (* bottom left rule *)
-        end
-
-    (* Restrict a judgement by two equations `ext0 = eps0` and `ext1 = eps1`. *)
-    fun Two (jdg : abt jdg) (ext0 : param) (eps0 : param) (ext1 : param) (eps1 : param) : abt jdg list =
-      if P.eq Sym.eq (ext0, eps0) then One jdg ext1 eps1 (* top right rule *)
-      else if P.eq Sym.eq (ext1, eps1) then One jdg ext0 eps0 (* top right rule *)
-      else if dimsContradictory (ext0, eps0) then [] (* second row, left rule *)
-      else if dimsContradictory (ext1, eps1) then [] (* second row, left rule *)
-      else
-        let
-          val (P.VAR u, P.APP _) = (ext0, eps0)
-          val (P.VAR v, P.APP _) = (ext1, eps1)
-        in
-          if Sym.eq (u, v) andalso not (P.eq Sym.eq (eps0, eps1)) then [] (* second row, right rule *)
-          else
-            [Seq.map (substSymbol (eps0, u)) (Seq.map (substSymbol (eps1, v)) jdg)] (* bottom right rule *)
-        end
+    (* Restrict a judgement (as the goal) by a list of equations. *)
+    val restrict : abt jdg -> (param * param) list -> abt jdg list
+  end
+  =
+  struct
+    (* A helper function which does substitution in a parameter. *)
+    fun substSymInParam (r, v) (P.APP t) = P.APP (P.Sig.map (substSymInParam (r, v)) t)
+      | substSymInParam (r, v) (P.VAR u) = if Sym.eq (u, v) then r else P.VAR u
```

Comment:
I think that this can be implemented as:

```
fun substSymInParam (r, v) t = P.bind (fn u => if Sym.eq (u, v) then r else P.VAR u)
```

or something along these lines. `ABT_PARAMETER_TERM` includes `MONAD`, which captures substitution.